tests/test_synthetic_volume.py

```diff
@@ -14,127 +14,120 @@
     TSymmetricVolume,
 )
 
-
-class Base:
-    def setUp(self):
-        self.dtype = np.float32
-        self.C = 1
-        self.seed = 0
-        vol_kwargs = dict(
-            L=self.L,
-            C=self.C,
-            seed=self.seed,
-            dtype=self.dtype,
-        )
-        if hasattr(self, "order"):
-            vol_kwargs["order"] = self.order
-
-        self.vol_obj = self.vol_class(**vol_kwargs)
-        self.vol = self.vol_obj.generate()
-
-    def testVolumeRepr(self):
-        """Test Synthetic Volume repr"""
-        self.assertTrue(
-            repr(self.vol_obj).startswith(f"{self.vol_obj.__class__.__name__}")
+# dtype fixture to pass into volume fixture.
+DTYPES = [np.float32, pytest.param(np.float64, marks=pytest.mark.expensive)]
+
+
+@pytest.fixture(params=DTYPES)
+def dtype_fixture(request):
+    dtype = request.param
+    return dtype
+
+
+# Parameter combinations for testing SyntheticVolumes with cyclic and dihedral symmetry.
+# Each tuple represents (volume class, resolution in pixels, cyclic order)
+PARAMS_Cn_Dn = [
+    (CnSymmetricVolume, 20, 2),
+    (CnSymmetricVolume, 21, 2),
+    (CnSymmetricVolume, 30, 3),
+    (CnSymmetricVolume, 31, 3),
+    (CnSymmetricVolume, 40, 4),
+    (CnSymmetricVolume, 41, 4),
+    (CnSymmetricVolume, 52, 5),
+    (CnSymmetricVolume, 53, 5),
+    (CnSymmetricVolume, 64, 6),
+    (CnSymmetricVolume, 65, 6),
+    (DnSymmetricVolume, 20, 2),
+    (DnSymmetricVolume, 21, 2),
+    (DnSymmetricVolume, 40, 3),
+    (DnSymmetricVolume, 41, 3),
+    (DnSymmetricVolume, 42, 4),
+    (DnSymmetricVolume, 43, 4),
+    (DnSymmetricVolume, 55, 5),
+    (DnSymmetricVolume, 56, 5),
+    (DnSymmetricVolume, 64, 6),
+    (DnSymmetricVolume, 65, 6),
+]
+
+
+# Parameters for tetrahedral, octahedral, asymmetric, and legacy volumes.
+# These volumes do not have an `order` parameter.
+VOL_CLASSES = [TSymmetricVolume, OSymmetricVolume, AsymmetricVolume, LegacyVolume]
+RESOLUTIONS = [20, 21]
+PARAMS = list(itertools.product(VOL_CLASSES, RESOLUTIONS))
+
+
+def vol_fixture_id(params):
+    vol_class = params[0]
+    res = params[1]
+    if len(params) > 2:
+        order = params[2]
+        return f"{vol_class.__name__}, res={res}, order={order}"
+    else:
+        return f"{vol_class.__name__}, res={res}"
+
+
+# Create SyntheticVolume fixture for the set of parameters.
+@pytest.fixture(params=PARAMS_Cn_Dn + PARAMS, ids=vol_fixture_id)
+def vol_fixture(request, dtype_fixture):
+    params = request.param
+    vol_class = params[0]
+    res = params[1]
+    vol_kwargs = dict(
+        L=res,
+        C=1,
+        seed=0,
+        dtype=dtype_fixture,
+    )
+    if len(params) > 2:
+        vol_kwargs["order"] = params[2]
+
+    return vol_class(**vol_kwargs)
+
+
+# SyntheticVolume tests:
+def test_volume_repr(vol_fixture):
+    """Test Synthetic Volume repr"""
+    assert repr(vol_fixture).startswith(f"{vol_fixture.__class__.__name__}")
```

Comment:
lets debug log it too

Reply:
Same as above, we also get a detailed message on the failed assert:

```
_______________________________________ test_volume_repr[LegacyVolume, res=21-float32] _______________________________________

vol_fixture = LegacyVolume {'order': 1, 'K': 16, 'alpha': 1.0, 'L': 21, 'C': 1, 'seed': 0, 'dtype': <class 'numpy.float32'>}

    def test_volume_repr(vol_fixture):
        """Test Synthetic Volume repr"""
>       assert repr(vol_fixture).startswith(f"bad stuff {vol_fixture.__class__.__name__}")
E       assert False
E        +  where False = <built-in method startswith of str object at 0x7f7c4df68850>('bad stuff LegacyVolume')
E        +    where <built-in method startswith of str object at 0x7f7c4df68850> = "LegacyVolume {'order': 1, 'K': 16, 'alpha': 1.0, 'L': 21, 'C': 1, 'seed': 0, 'dtype': <class 'numpy.float32'>}".startswith
E        +      where "LegacyVolume {'order': 1, 'K': 16, 'alpha': 1.0, 'L': 21, 'C': 1, 'seed': 0, 'dtype': <class 'numpy.float32'>}" = repr(LegacyVolume {'order': 1, 'K': 16, 'alpha': 1.0, 'L': 21, 'C': 1, 'seed': 0, 'dtype': <class 'numpy.float32'>})

tests/test_synthetic_volume.py:94: AssertionError
```

Reply:
When it passes, do you know what the repr is from the logs?

Reply:
Nope. I've added a debug log here.

Reply:
Thanks :)